rfd/001-tcp-load-balancer-design-doc.md

```diff
@@ -0,0 +1,119 @@
+
+Our use-case follows the #2 reason for leveraging `sync.Map` over a standard map with a Mutex; our disjoint sets of keys will be the unique client ids.
+
+
+### Upstream Authorization Management
+
+The load balancer will employ a simple authorization scheme which will deny access to specific upstreams for certain clients. This scheme will be statically defined in code. Once the load balancer has determined that the client is authenticated and within client connection limit (and prior to selecting the least connections host) it will filter out unauthorized hosts for that client. To simply demonstrate this capability, we store a map of `clientID`s (based on client certificate) to disallowed `hostID`s; because host configuration is outside the scope of this project, the order in which the host was registered with the load balancer will be used as the `hostID`.
+
+### Request Forwarder
+
+Incoming packets will be forwarded after checking 3 criterion:
+* Client must be authenticated.
+* Client must be within connection limit.
+* Client must be authorized to access at least one healthy host.
+
+Once these requirements are satisfied, the request will be forwarded to an authorized host with the least connections.
+
+The least connection tracking will prioritize simplicity over performance for this project. Each new connection that is made will increment a host's connection count, and each completed connection will decrement the host's connection count. Each incoming request will iterate through all authorized hosts to find the one with the least active connections. A more performant solution could later be devised which may not have to iterate over every authorized host for every request (likely leveraging an async queue). One additional shortcoming of this simple approach is that a large batch of requests which arrive at exact same time could be routed to the exact same host, but this initial project is not planning to handle that edge case.
+
+Example of simple approach for selecting host:
+```
+func (l *LoadBalancer) LeastConnections() *upstream.TcpHost {
+	if l == nil || len(l.hosts) == 0 {
+		return nil
+	}
+
+	host := l.hosts[0]
+
+	for i, h := range l.hosts[1:len(hosts)] {
+		if h.ConnectionCount() < host.ConnectionCount() {
+			host = h
+		}
+	}
+
+	return host
+}
+```
+
+Once the host is selected, data will be passed directly from clients to upstream hosts, without being stored on the load balancer.
+
+### Health Checks
+The load balancer removes unhealthy upstreams if it is unable to connect to the host while handling a client connection. The health checks could include a grace period, or some request-forwarding backoff, allowing for n attempts before removal, but for simplicity, this project will immediately remove the host from available hosts upon a single failed connection.
```

Comment:
> immediately remove the host from available hosts upon a single failed connection.

How removed host will be restored ?

Reply:
Hosts could be restored in a few ways. Here's a couple ideas:

1. Place the responsibility on hosts to re-register when healthy (which requires a non-static approach to host registration -- not yet planned to be in scope).
2. Persist registry of all registered hosts and store health status for each host. Spin off separate process to periodically recheck unhealthy hosts so that statuses can be updated.

To minimize scope on this project, I had planned to statically define hosts and not to restore disconnected hosts if/when they regain health. Is that approach reasonable, or should we include host restoration within scope?

Reply:
Okay, I'll add option 2 to the design doc. 

A health checking goroutine will be satisfactory.